### ISSUE TITLE
Is it possible to call a proc on `throttled? == true`

### ISSUE THREAD

**Issue by bramswenson** [ixti/sidekiq-throttled#47]
I'd like to do something simple like `Airbrake.notify` whenever jobs are throttled. Is there a way to add a callback type behavior? I've looked through the code and don't see anything native off hand. TIA!

**Comment by ixti** [comment]
It's not available ATM, but I will think on adding this feature - I like it.

**Comment by bramswenson** [comment]
@ixti I'd be happy to submit a PR with a little guidance on how you'd like to see this implemented. Thanks!

**Comment by ixti** [comment]
I guess we can come with something as simple as:

``` ruby
class MyJob
  include Sidekiq::Worker
  include Sidekiq::Throttled::Worker

  OBSERVER = -> (strategy) do
    Rails.logger.warn("MyJob was throttled by #{strategy}")
  end

  sidekiq_throttle({
    :concurrency => { :limit => 42 },
    :observe => OBSERVER
  })
end
```

**Comment by ixti** [comment]
Idea is pretty same to `:key` option, so that if it's on top-level it's passed to both strategies.

**Comment by ixti** [comment]
Or in fact it might be even simplier - and remain top-level only. So that in here:

https://github.com/sensortower/sidekiq-throttled/blob/master/lib/sidekiq/throttled/strategy.rb#L57-L66

you will just add something like:

``` ruby
if @concurrency && @concurrency.throttled?(...)
  @observer&.call(@concurrency)
  return true
end

 # ...
```

**Comment by ixti** [comment]
In fact it might be implemented per-strategy - will be even easier.

**Comment by bramswenson** [comment]
Thanks, I'll look into this!

**Comment by ogins57** [comment]
I think it's better to add callback process which will be called **after throttled** per threshold with `:after_throttled` option:

``` ruby
class MyWorker
  include Sidekiq::Worker
  include Sidekiq::Throttled::Worker

  sidekiq_options :queue => :my_queue

  # Define a callback process
  AFTER_THROTTLED = -> do
    puts "MyJob was throttled"
  end

  sidekiq_throttle({
    :concurrency => { :limit => 10 },
    # Apply a callback process
    :threshold => { :limit => 1_000, :period => 1.hour, after_throttled: AFTER_THROTTLED }
  })

  def perform
    # ...
  end
end
```

There are several points:

1. The callback process will be basically called **ONLY ONCE** after throttled in terms of impact on performance.
2. The callback process doesn't have any arguments.
3. Doesn't provide callback function in `concurrency` because some settings will be throttled so frequently. (e.g. `concurrency { limit: 1 }` will cause throttling on every sidekiq polling )

As for No.2, I first assumed that it is better to process `*job_args` or somthing like key_suffix, however, given No.1 (called ONLY ONCE) and hence it can't handle every time throttled, I made the conclusion not to have any arguments.

**Comment by ogins57** [comment]
>1. The callback process will be basically called **ONLY ONCE** after throttled in terms of impact on performance.

This description was not clear, I meant that frequently processing should be prevented in terms of impact on performance and don't mean will be never processed in the future. I think it is better to prevent during a period of ttl from the throttling started, like below.

``` ruby
def throttled?(*job_args)
  return false unless (job_limit = limit(job_args))

  key = key(job_args)
  period = period(job_args)

  kwargs = {
    :keys => [key],
    :argv => [job_limit, period, Time.now.to_f]
  }

  throttled = Sidekiq.redis { |redis| 1 == SCRIPT.eval(redis, kwargs) }
  callback_once("#{key}:callback", period) if throttled
  throttled
end

def callback_once(key, ttl)
  MUTEX.synchronize do
    return if callback_paused?(key)

    @callback&.call
    pause_callback(key, ttl)
  end
end

def pause_callback(key, ttl)
  Sidekiq.redis { |conn| conn.setex(key, ttl.to_i, "pause") }
end

def callback_paused?(key)
  Sidekiq.redis { |conn| conn.exists(key) }
end
```

**Comment by ixti** [comment]
> I think it's better to add callback process which will be called after throttled per threshold with :after_throttled option:

The idea is to provide a generic observer. For both types of throttling. Also I really don't like `after_*` name. It sounds as if there's corresponding `before_*` and `around_*`. While in our case we just need to notify of some event happened.

> The callback process doesn't have any arguments.

No. We want to pass arguments that caused throttling. That's essential. For example, let's say there's a job that pings different hosts based on given URL, on throttling we would really be curious which host throttles us most.

> Doesn't provide callback function in concurrency because some settings will be throttled so frequently. (e.g. concurrency { limit: 1 } will cause throttling on every sidekiq polling )

I don't understand reasoning. Configuration of concurrency is NOT fixed, and in different cases it might be different. And we definitely want to know when concurrency occurs if we don't expect that. Your reason of not observing concurrency throttling can be easily applied to threshold as well - user can set VERY low limit of threshold just as easy as he can do that with concurrency - both strategies are configurable and can occur on every sidekiq polling.

> The callback process will be basically called ONLY ONCE after throttled in terms of impact on performance. I meant that frequently processing should be prevented in terms of impact on performance and don't mean will be never processed in the future. I think it is better to prevent during a period of ttl from the throttling started, like below.

That is totally out of the scope of this feature or sidekiq-throttled in general. It's consumer's responsibility to throttle/skip notifications not our. Sidekiq-throttled should be simple and straight forward in this aspects. Skipping notifications on any conditions is not what we should do. Observer should do that on it's own.

---

### API design

I see 2 ways of this feature to be implemented. First is straight forward and I want this iteration of the feature to focus exactly on this version:

``` ruby
class MyWorker
  # ...

  MY_OBSERVER = lambda do |strategy, url|
    return if :concurrency  == strategy # skip concurrency if we want that

    App.statsd.increment "my_worker:#{URI(url).host}"
  end 
 
  sidekiq_throttled **some_config, :observe => MY_OBSERVER

  def perform(url)
    # ...
  end
end
```

In other words, observer is an object that responds to `#call` message, and accepts 1+ arguments (first argument is strategy name, rest is what was passed to `#perform`.

Second way is to provide `Sidekiq::Observable` mixin that will enhance child classes with `.on` and `.notify` methods. But conceptually under the hood both ways will look pretty much the same.

See Also: https://en.wikipedia.org/wiki/Observer_pattern


### Performance

If you want to make this more performant, we can discuss this after the first iteration of this feature. As I see how we can make notifications to be handled in a dedicated thread. But most likely this is totally out of the sidekiq-throttled scope.

**Comment by ogins57** [comment]
Thank you for sharing your opinion. I totally agree with you.

>Doesn't provide callback function in concurrency because some settings will be throttled so frequently. (e.g. concurrency { limit: 1 } will cause throttling on every sidekiq polling )

I first thought that there was `:period` option in threshold which means 'threshold is controllable' unlike concurrency, but depending on the configuration is the same as concurrency and as you said there will be definitely the case we want to know when concurrency occurs. Thanks!

>### Performance

Aside from performance, if neither using thresholds' period as TTL nor applying option such as a `:observe_duration`, MY_OBSERVER will be called the same number of sidekiq workers times on every sidekiq polling once after reached to limit. I think it seems a bit annoying but I understand that it's out of the sidekiq-throttled scope, so I won't care it in the first iteration of this feature.

**Comment by ixti** [comment]
Implemented as of v0.11.0
